llvm/lib/CodeGen/GlobalISel/CombinerHelperCasts.cpp

```diff
@@ -273,3 +273,42 @@ bool CombinerHelper::matchExtOfExt(const MachineInstr &FirstMI,
 
   return false;
 }
+
+bool CombinerHelper::matchCastOfBuildVector(const MachineInstr &CastMI,
+                                            const MachineInstr &BVMI,
+                                            BuildFnTy &MatchInfo) {
+  const GExtOrTruncOp *Cast = cast<GExtOrTruncOp>(&CastMI);
+  const GBuildVector *BV = cast<GBuildVector>(&BVMI);
+
+  if (!MRI.hasOneNonDBGUse(BV->getReg(0)))
+    return false;
+
+  Register Dst = Cast->getReg(0);
+  // The type of the new build vector.
+  LLT DstTy = MRI.getType(Dst);
+  // The scalar or element type of the new build vector.
+  LLT ElemTy = DstTy.getScalarType();
+  // The scalar or element type of the old build vector.
+  LLT InputElemTy = MRI.getType(BV->getReg(0)).getScalarType();
```

Comment:
You know the source is a vector, so just use getElementType, not getScalarType. Better yet, just take the type of a source operand? Just in case we go back to having the implicit cast in build_vector like the dag

Reply:
G_BUILD_VECTOR has no casting. We have an explicit G_BUILD_VECTOR_TRUNC.